src/psyclone/parse/module_info.py

```diff
@@ -135,63 +146,53 @@ def get_source_code(self):
 
         '''
         try:
-            return self._file_info.contents
+            return self._file_info.get_source_code()
         except FileNotFoundError as err:
             raise ModuleInfoError(
                 f"Could not find file '{self._file_info.filename}' when trying"
                 f" to read source code for module '{self._name}'") from err
 
     # ------------------------------------------------------------------------
-    def get_parse_tree(self):
-        '''Returns the fparser AST for this module. The first time, the file
-        will be parsed by fparser using the Fortran 2008 standard. The AST is
-        then cached for any future uses.
+    def get_fparser_tree(self) -> Union[Program, None]:
+        '''Returns the fparser AST for this module.
 
-        :returns: the fparser AST for this module.
-        :rtype: :py:class:`fparser.two.Fortran2003.Program`
+        :returns: The fparser AST for this module.
 
+        :raises ModuleInfoError: When errors have been triggered while
+            creating the fparser tree.
         '''
-        if not self._parse_attempted:
-            # This way we avoid that any other function might trigger to
-            # parse this file again (in case of parsing errors).
-            self._parse_attempted = True
-
-            reader = FortranStringReader(
-                self.get_source_code(),
-                include_dirs=Config.get().include_paths)
-            parser = ParserFactory().create(std="f2008")
-            self._parse_tree = parser(reader)
+        try:
+            return self._file_info.get_fparser_tree()
```

Comment:
Does this retain the feature that we will not end up repeatedly attempting to parse some code that fparser fails on? I don't think it does. Please could you extend FileInfo to keep track of this.

Reply:
Indeed. This wansn't the case, yet.
I added this.
Thanks for spotting this performance bug!